clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -1458,6 +1512,243 @@ inline void csr2csc(sycl::queue queue, int m, int n, int nnz,
+/// numbering.
+/// \param [in] info_b Matrix info of the matrix B.
+/// \param [in] val_b An array containing the non-zero elements of the matrix B.
+/// \param [in] row_ptr_b An array of length row number + 1.
+/// \param [in] col_ind_b An array containing the column indices in index-based
+/// numbering.
+/// \param [in] info_c Matrix info of the matrix C.
+/// \param [out] val_c An array containing the non-zero elements of the matrix
+/// C.
+/// \param [in] row_ptr_c An array of length row number + 1.
+/// \param [out] col_ind_c An array containing the column indices in index-based
+/// numbering.
+template <typename T>
+void csrgemm(descriptor_ptr desc, oneapi::mkl::transpose trans_a,
+             oneapi::mkl::transpose trans_b, int m, int n, int k,
+             const std::shared_ptr<matrix_info> info_a, const T *val_a,
+             const int *row_ptr_a, const int *col_ind_a,
+             const std::shared_ptr<matrix_info> info_b, const T *val_b,
+             const int *row_ptr_b, const int *col_ind_b,
+             const std::shared_ptr<matrix_info> info_c, T *val_c,
+             const int *row_ptr_c, int *col_ind_c) {
+  using Ty = typename ::dpct::detail::lib_data_traits_t<T>;
+  sycl::queue queue = desc->get_queue();
+  detail::csrgemm_args_info args(trans_a, trans_b, m, n, k, info_a, val_a,
+                                 row_ptr_a, col_ind_a, info_b, val_b, row_ptr_b,
+                                 col_ind_b, info_c, row_ptr_c);
+  const auto Iter = desc->get_csrgemm_info_map().find(args);
+  if (Iter == desc->get_csrgemm_info_map().end()) {
+    throw std::runtime_error("csrgemm_nnz is not invoked previously.");
+  }
+  oneapi::mkl::sparse::matrix_handle_t *matrix_handle_a =
+      Iter->second.matrix_handle_a;
+  oneapi::mkl::sparse::matrix_handle_t *matrix_handle_b =
+      Iter->second.matrix_handle_b;
+  oneapi::mkl::sparse::matrix_handle_t *matrix_handle_c =
+      Iter->second.matrix_handle_c;
+  oneapi::mkl::sparse::matmat_descr_t *matmat_desc = Iter->second.matmat_desc;
+
+  auto data_col_ind_c = dpct::detail::get_memory<int>(col_ind_c);
+  auto data_val_c = dpct::detail::get_memory<Ty>(val_c);
+  auto data_row_ptr_c = dpct::detail::get_memory<int>(row_ptr_c);
+  int rows_c = m;
+  int cols_c = n;
+  oneapi::mkl::sparse::set_csr_data(queue, *matrix_handle_c, rows_c, cols_c,
+                                    info_c->get_index_base(), data_row_ptr_c,
+                                    data_col_ind_c, data_val_c);
+  sycl::event e;
+#ifndef DPCT_USM_LEVEL_NONE
+  e =
+#endif
+      __MATMAT(oneapi::mkl::sparse::matmat_request::finalize, nullptr);
+#undef __MATMAT
+
+  std::vector<sycl::event> e_s;
+  e_s.push_back(
+      oneapi::mkl::sparse::release_matrix_handle(queue, matrix_handle_a, {e}));
+  e_s.push_back(
+      oneapi::mkl::sparse::release_matrix_handle(queue, matrix_handle_b, {e}));
+  e_s.push_back(
+      oneapi::mkl::sparse::release_matrix_handle(queue, matrix_handle_c, {e}));
```

Comment:
maybe we can unify the management of matrix_handle and matmat_desc (use a base class)